gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java

```diff
@@ -0,0 +1,357 @@
+/*
+ * Copyright (C) 2014-2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.util;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.avro.AvroRuntimeException;
+import org.apache.avro.Schema;
+import org.apache.commons.lang3.StringUtils;
+import org.codehaus.jackson.JsonNode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Preconditions;
+
+public class AvroFlattener {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AvroFlattener.class);
+
+  /***
+   * Flatten the Schema to un-nest recursive Records (to make it optimal for ORC)
+   * @param schema Avro Schema to flatten
+   * @return Flattened Avro Schema
+   */
+  public static Schema flatten(Schema schema) {
+    Preconditions.checkNotNull(schema);
+
+    Schema flattenedSchema = flatten(schema, false);
+
+    LOG.debug("Original Schema : " + schema);
+    LOG.debug("Flattened Schema: " + flattenedSchema);
+
+    return flattenedSchema;
+  }
+
+  private static Schema flatten(Schema schema, boolean shouldPopulateLineage) {
+    Schema flattenedSchema;
+
+    // Process all Schema Types
+    // (Primitives are simply cloned)
+    switch (schema.getType()) {
+      case ARRAY:
+        // Array might be an array of recursive Records, flatten them
+        flattenedSchema = Schema.createArray(flatten(schema.getElementType(), false));
```

Comment:
I will add some comments, but in quick summary: shouldPopulateLineage is set to true if the field is going to be flattened and moved up the hierarchy - so that lineage information can be tagged to it. Which happens when there is Record within Record OR Record within Option with Record and so on, however not when there is Record within Map within Record